onnxruntime/core/platform/windows/logging/etw_sink.cc

```diff
@@ -137,28 +137,47 @@ void NTAPI EtwRegistrationManager::ORT_TL_EtwEnableCallback(
 EtwRegistrationManager::~EtwRegistrationManager() {
   std::lock_guard<OrtMutex> lock(callbacks_mutex_);
   callbacks_.clear();
-  ::TraceLoggingUnregister(etw_provider_handle);
+  if (initialized_ || initializing_) {
+    std::lock_guard<OrtMutex> init_lock(init_mutex_);
+      assert(!initializing_);
+      if (initialized_) {
+        ::TraceLoggingUnregister(etw_provider_handle);
+        initialized_ = false;
+      }
+  }
 }
 
 EtwRegistrationManager::EtwRegistrationManager() {
 }
 
-void EtwRegistrationManager::LazyInitialize() {
-  if (!initialized_) {
+void EtwRegistrationManager::LazyInitialize() try {
+  if (!initialized_ && !initializing_) {
     std::lock_guard<OrtMutex> lock(init_mutex_);
-    if (!initialized_) {  // Double-check locking pattern
-      initialized_ = true;
+    if (!initialized_ && !initializing_) {  // Double-check locking pattern
+      initializing_ = true;
       etw_status_ = ::TraceLoggingRegisterEx(etw_provider_handle, ORT_TL_EtwEnableCallback, nullptr);
       if (FAILED(etw_status_)) {
         ORT_THROW("ETW registration failed. Logging will be broken: " + std::to_string(etw_status_));
       }
+      initialized_ = true;
+      initializing_ = false;
     }
   }
+} catch (...)
+{
+  initialized_ = false;
+  initializing_ = false;
```

Comment:
In theory they are mutually exclusive, so maybe a `State` enum would be cleaner (with options `NotInitialized, Initializing, Initialized`). But it's not a big deal.